src/neo/SmartContract/ApplicationEngine.Contract.cs

```diff
@@ -62,44 +68,39 @@ protected internal void CreateContract(byte[] script, byte[] manifest)
                 CallContractInternal(contract, md, new Array(ReferenceCounter) { false }, CallFlags.All, CheckReturnType.EnsureIsEmpty);
         }
 
-        protected internal void UpdateContract(byte[] script, byte[] manifest)
+        protected internal void UpdateContract(byte[] nefFile, byte[] manifest)
         {
-            if (script is null && manifest is null) throw new ArgumentException();
+            if (nefFile is null && manifest is null) throw new ArgumentException();
 
-            AddGas(StoragePrice * ((script?.Length ?? 0) + (manifest?.Length ?? 0)));
+            AddGas(StoragePrice * ((nefFile?.Length ?? 0) + (manifest?.Length ?? 0)));
 
-            var contract = Snapshot.Contracts.TryGet(CurrentScriptHash);
+            var contract = Snapshot.Contracts.GetAndChange(CurrentScriptHash);
             if (contract is null) throw new InvalidOperationException($"Updating Contract Does Not Exist: {CurrentScriptHash}");
 
-            if (script != null)
+            if (nefFile != null)
             {
-                if (script.Length == 0 || script.Length > MaxContractLength)
-                    throw new ArgumentException($"Invalid Script Length: {script.Length}");
-                UInt160 hash_new = script.ToScriptHash();
-                if (hash_new.Equals(CurrentScriptHash) || Snapshot.Contracts.TryGet(hash_new) != null)
-                    throw new InvalidOperationException($"Adding Contract Hash Already Exist: {hash_new}");
-                contract = new ContractState
-                {
-                    Id = contract.Id,
-                    Script = script.ToArray(),
-                    Manifest = contract.Manifest
-                };
-                contract.Manifest.Abi.Hash = hash_new;
-                Snapshot.Contracts.Add(hash_new, contract);
-                Snapshot.Contracts.Delete(CurrentScriptHash);
+                if (nefFile.Length == 0 || nefFile.Length > NefFile.MaxLength)
+                    throw new ArgumentException($"Invalid NefFile Length: {nefFile.Length}");
+
+                NefFile nef = nefFile.AsSerializable<NefFile>();
+                if (nef.Script.Length == 0 || nef.Script.Length > MaxContractLength)
+                    throw new ArgumentException($"Invalid Script Length: {nef.Script.Length}");
+
+                // Update script
+                contract.Script = nef.Script;
             }
             if (manifest != null)
             {
                 if (manifest.Length == 0 || manifest.Length > ContractManifest.MaxLength)
                     throw new ArgumentException($"Invalid Manifest Length: {manifest.Length}");
-                contract = Snapshot.Contracts.GetAndChange(contract.ScriptHash);
                 contract.Manifest = ContractManifest.Parse(manifest);
                 if (!contract.Manifest.IsValid(contract.ScriptHash))
```

Comment:
Remove this constraint?

Reply:
But this also check if the new `Groups` are well signed